src/prefect/server/services/task_run_recorder.py

```diff
@@ -91,91 +103,170 @@ def task_run_from_event(event: ReceivedEvent) -> TaskRun:
+        for event in events
+        for task_run, task_run_dict in [db_recordable_task_run_from_event(event)]
+    ]
+
+    # Batch by state type to avoid column mismatches during bulk insert
+    batches_by_state: dict[StateType, list] = {}
+    for tr in all_task_runs:
+        state_type = tr["task_run"].state.type
+        batches_by_state.setdefault(state_type, []).append(tr)
+
+    state_type_counts = {
+        state_type: len(batch) for state_type, batch in batches_by_state.items()
+    }
+    logger.debug(
+        f"Partitioned task runs into {len(batches_by_state)} groups by state type for bulk insertion: {state_type_counts}"
+    )
+
+    db = provide_database_interface()
+
+    async with db.session_context() as session:
+        for state_type in StateType:
+            if state_type not in batches_by_state:
+                continue
+
+            batch = batches_by_state[state_type]
+
+            # Should be the same for all in the batch
+            update_cols = batch[0]["task_run_dict"].keys()
 
-        db = provide_database_interface()
-        async with db.session_context() as session:
-            # Combine all attributes for a single atomic operation
-            all_attributes = {
-                **task_run_attributes,
-                **denormalized_state_attributes,
-                "created": prefect.types._datetime.now("UTC"),
-            }
-
-            # Single atomic INSERT ... ON CONFLICT DO UPDATE
-            await session.execute(
-                db.queries.insert(db.TaskRun)
-                .values(**all_attributes)
-                .on_conflict_do_update(
-                    index_elements=["id"],
-                    set_={
-                        "updated": prefect.types._datetime.now("UTC"),
-                        **task_run_attributes,
-                        **denormalized_state_attributes,
+            logger.debug(
+                f"Preparing to bulk insert {len(batch)} task runs with state type {state_type}"
+            )
+            to_insert = [
+                tr["task_run_dict"] | {"created": now, "updated": now} for tr in batch
+            ]
+
+            insert_statement = db.queries.insert(db.TaskRun).values(to_insert)
+            upsert_statement = insert_statement.on_conflict_do_update(
+                index_elements=["id"],
+                set_={
+                    # See https://www.postgresql.org/docs/current/sql-insert.html for details on excluded.
+                    # Idea is excluded.x references the proposed insertion value for column x.
+                    **{
+                        col.name: getattr(insert_statement.excluded, col.name)
+                        for col in insert_statement.excluded
+                        if col.name in update_cols
                     },
-                    where=db.TaskRun.state_timestamp < task_run.state.timestamp,
-                )
+                },
```

Comment:
I'm pretty sure we need to include `updated` here to ensure that it is written on upsert since it won't be included in `update_cols`.
```suggestion
                set_={
                    # See https://www.postgresql.org/docs/current/sql-insert.html for details on excluded.
                    # Idea is excluded.x references the proposed insertion value for column x.
                    **{
                        col.name: getattr(insert_statement.excluded, col.name)
                        for col in insert_statement.excluded
                        if col.name in update_cols
                    },
                    "updated": now,
},
```

Reply:
Okay, fixed this and included a specific unit test to cover it.